src/EFCore.Cosmos/Infrastructure/Internal/CosmosModelValidator.cs

```
@@ -313,30 +320,33 @@ public override void Validate(IModel model, IDiagnosticsLogger<DbLoggerCategory.
                 throw new InvalidOperationException(CosmosStrings.NoIdKey(entityType.DisplayName(), idProperty.Name));
             }
 
-            var partitionKeyPropertyName = entityType.GetPartitionKeyPropertyName();
-            if (partitionKeyPropertyName != null)
+            var partitionKeyPropertyNames = entityType.GetPartitionKeyPropertyNames();
+            if (partitionKeyPropertyNames.Count == 0)
             {
-                var partitionKey = entityType.FindProperty(partitionKeyPropertyName);
-                if (partitionKey == null)
-                {
-                    throw new InvalidOperationException(
-                        CosmosStrings.PartitionKeyMissingProperty(entityType.DisplayName(), partitionKeyPropertyName));
-                }
-
-                var partitionKeyType = partitionKey.GetTypeMapping().Converter?.ProviderClrType
-                    ?? partitionKey.ClrType;
-                if (partitionKeyType != typeof(string))
+                logger.NoPartitionKeyDefined(entityType);
+            }
+            else if (partitionKeyPropertyNames.Count > 3)
+            {
+                throw new InvalidOperationException(
+                    CosmosStrings.PartitionKeyTooManyParts(entityType.DisplayName(), string.Join(',', partitionKeyPropertyNames)));
+            }
+            else
+            {
+                foreach (var partitionKeyPropertyName in partitionKeyPropertyNames)
                 {
-                    throw new InvalidOperationException(
-                        CosmosStrings.PartitionKeyNonStringStoreType(
-                            partitionKeyPropertyName, entityType.DisplayName(), partitionKeyType.ShortDisplayName()));
```

We should still preemptively throw if it's an unsupported type

---

I've added this back in, but I don't know how to test it. As far as I know, any type mapping will be converted to one of these types (because these are the types that are supported by Cosmos) or will not map at all because the type is not supported by Cosmos. Is there a type Cosmos supports that can't be used here?

---

You can use `JObject`

---

JObject doesn't work. The partition key is made part of the key, so this is thrown before the validation code for the partition key type is reached:

```
Property 'Customer.__foo' cannot be used as a key because it has type 'JObject' which does not implement 'IComparable<T>', 'IComparable' or 'IStructuralComparable'. Use 'HasConversion' in 'OnModelCreating' to wrap 'JObject' with a type that can be compared.
```

---

Ok, then you can add a test TypeMappingSource that allows any type.